### PR TITLE
refactor: rewrite `createRule` and `RuleTester`

### PR DESCRIPTION
**Pull request by SukkaW**
Currently, `eslint-plugin-import-x` patches `@typescript-eslint/utils` to add custom properties like the `category` and modify properties like the `recommended`.

The PR accomplishes the same thing, by doing some TypeScript exercises instead.

### CODE REVIEW

**Comment by SukkaW** [comment]
@JounQin @silverwind What do you think? Do you prefer creating our own utilities over patching `ts-eslint`?

File: src/utils/create-rule.ts [hunk 1]
```diff
@@ -1,5 +1,70 @@
...
+      ...rules,
+    }),
+    meta: {
+      ...meta,
+      docs: {
+        ...docs,
+        url: docsUrl(name, commitHash),
+      },
+    },
+  }
```
Comment (SukkaW) [inline src/utils/create-rule.ts:69]: Under the hood we are still using `ts-eslint` as much as possible, we are only overriding the returned `meta` property with our `meta` property afterward.

File: test/utils.ts [hunk 2]
```diff
@@ -164,3 +166,45 @@ export const SYNTAX_CASES = [
...
+
+    const modifiedRule: Parameters<TSESLint.RuleTester['run']>[1] = {
+      ...ruleWithoutMeta,
+      meta: {
+        ...metaWithoutDocs,
+        docs: docsWithoutCategoryAndRecommend,
+      },
+    }
+
+    return super.run(ruleName, modifiedRule, tests)
```
Comment (SukkaW) [inline test/utils.ts:208]: Under the hood, `run$` does some extra modification before `super.run`.

File: test/utils.ts [hunk 3]
```diff
@@ -164,3 +166,45 @@ export const SYNTAX_CASES = [
...
+
+export class RuleTester extends TSESLint.RuleTester {
+  constructor(option: RuleTesterOptions = {}) {
+    /**
+     * The parser option is required in TSESLint.RuleTester, but not in ESLint.RuleTester.
+     * Since TSESLint.RuleTester is just an alias of ESLint.RuleTester, we can safely ignore the type error.
+     */
+    // @ts-expect-error -- parser is not required
+    super(option)
+  }
```
Comment (SukkaW) [inline test/utils.ts:182]: By extending our own `RuleTester` class, we don't have to patch the type of the argument.

File: test/utils.ts (hunk 2 above)
Comment (JounQin) [reply test/utils.ts:208]: Is that possible to use `// @ts-expect-error` here instead of create a new `run$` method?

Comment (SukkaW) [reply test/utils.ts:208]: @JounQin 

Adding a `ts-expect-error` and moving on sounds very tempting, but not a best practice here.

TypeScript throws a type mismatch for a reason: [Inheritance implies subtyping](https://stackoverflow.com/questions/23592131/what-is-the-difference-between-subtyping-and-inheritance-in-oo-programming), and subtype mismatch breaks the Liskov substitution. So TypeScript disallows inheriting from a base class while not being a valid subtype of it.

**Comment by SukkaW** [comment]
@JounQin Would you like to cleat the cache of the GitHub Action? It seems that the CI failed to the corrupt cache.

File: test/utils.ts (hunk 2 above)
Comment (JounQin) [reply test/utils.ts:208]: Hmm... I think it's some kind of balance choice here, or we can say that the typings in `tseslint` is incorrect? The modifications are quite huge, aren't they? I'm thinking about the maintenance vs the patch one.

**Comment by JounQin** [comment]
@SukkaW 

I raised https://github.com/typescript-eslint/typescript-eslint/issues/8897 to discuss about the optional `parser` property.

And I notice patches are also used in tseslint's repository https://github.com/typescript-eslint/typescript-eslint/blob/main/.yarn/patches, and considering there are some `// @ts-expect-error - flow type` comments in source codes, I'm also considering to add more patches for typings maybe, and the maintenance for such small patches should be easier than TypeScript exercises.

HDYT?

**Comment by SukkaW** [comment]
> HDYT?

@JounQin Actually, I kinda hate the `typescript-eslint` project. Their maintainers have little knowledge of TypeScript, yet they ignore most of the communities' suggestions. 

In fact, in my own `eslint-plugin-sukka`, I am writing my own `createRule` and `RuleTester`, without depending on either `eslint` or `typescript-eslint`.

File: test/utils.ts (hunk 2 above)
Comment (silverwind) [reply test/utils.ts:208]: Sorry, this is quite far over my head. Only feedback I would have is `run$` looks a bit dirty to me. I'd suggest some camelCase name instead.

**Comment by SukkaW** [comment]
I am closing the PR now since TypeScript ESLint 8 has addressed the custom property of `createRule` through generic parameters. The issue has already been resolved by me in another PR https://github.com/un-ts/eslint-plugin-import-x/pull/112.

**Pull request closed without merging**